test/map.hashmap.spec.js

```diff
@@ -0,0 +1,71 @@
+/* globals it */
+const assert = require('assert')
+const tsame = require('tsame')
+const Block = require('@ipld/block')
+const HashMap = require('../src/maps/hashmap')
+const { Lookup, get } = require('../')
+
+const same = (...args) => assert.ok(tsame(...args))
+const test = it
+
+const lookup = new Lookup()
+lookup.register(HashMap._type, HashMap)
+
+function storage () {
+  let kv = new Map()
+  return {
+    set (cid, value) {
+      return kv.set(cid.toString(), value)
+    },
+    get (cid) {
+      return kv.get(cid.toString())
+    },
+    lookup // so we can pass to get()
+  }
+}
+
+test('HashMap single-block set/get', async () => {
+  let store = storage()
+  let externalFixture = { some: 'other block' }
+  let external = Block.encoder(externalFixture, 'dag-json')
+  let externalCid = await external.cid()
+  store.set(externalCid, external)
+
+  // default IAMap options are a bitWidth of 5 and bucketSize of 8 giving 256 elements per block
+  // with a good hash and a few items, we're unlikely to overflow beyond the root block
+  let map = await HashMap.create(store, 'dag-json')
+  map = await map.set('foo', 'bar')
+  map = await map.set('bar', 100)
+  map = await map.set('baz', { wut: { is: 'this' } })
+  map = await map.set('boom', externalCid)
+
+  let root = map.id
+
+  same((await get(store, root, 'foo')).toString(), 'bar')
+  same((await get(store, root, 'bar')).toInt(), 100)
+  /* TODO: we should be able to navigate into child properties of values
```

Comment:
here are the parts where it seems that the path should allow us to navigate into objects stored inline in map values and transparently across links to other blocks where the value is a CID

Reply:
I’ll implement a `path` method does just that. I’d like the `get` method to stay close to the underlying type method but we should have a recursive version for full path traversal.

Reply:
Actually, forget everything I’ve said up until now :)

https://github.com/ipld/js-types/blob/master/src/kinds.js#L14

We should already do this ;) The type implementation shouldn’t return a `remaining` path, it should return a `get` for the remaining path if it can’t fulfill the entire path locally. I forgot that I made this upgrade in the wee hours.

It makes sense, this reduces the work load quite a bit. If someone wants to do a more incremental path resolution they can just break up the path themselves and resolve each part one at a time. The default behavior should be the most efficient, which is to pass along the entire path and resolve it entirely.

Reply:
It’s probably time to document this since I can’t even keep it all straight myself.

Reply:
I had `args{}` wrong, got it working now and it's pretty darn sweet